Do one instead of two heap allocations

rschu1ze opened the pull request:
make_shared<>() allocates the control block + the pointed-to-object in
one go, shared_ptr<T>() does two allocations. Therefore, now using the
former for shared_ptr_helper. (The downside is that with make_shared<>()
the shared_ptr will stick around if the weak count is > 0 but weak ptrs
are rarely used and minimally higher memory consumption should not be a
big problem.)

While at it, also made trait "is_shared_ptr" slightly more idiomatic.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Commit:
Do one instead of two heap allocations

make_shared<>() allocates the control block + the pointed-to-object in
one go, shared_ptr<T>() does two allocations. Therefore, now using the
former for shared_ptr_helper. (The downside is that with make_shared<>()
the shared_ptr will stick around if the weak count is > 0 but weak ptrs
are rarely used and minimally higher memory consumption should not be a
big problem.)

While at it, also made trait "is_shared_ptr" slightly more idiomatic.

antonio2368 reviewed (approved):
> The downside is that with make_shared<>()
the shared_ptr will stick around if the weak count is > 0 but weak ptrs
are rarely used and minimally higher memory consumption should not be a
big problem.)

I agree with this, `std::weak_ptr` is not used a lot, and it just increases the lifetime of object slightly.
I didn't know about the allocation difference, cool stuff!

antonio2368 commented on base/base/shared_ptr_helper.h:
```diff
@@ -11,25 +12,33 @@
 {
+private:
+    // https://stackoverflow.com/a/25069711
+    struct make_shared_enabler : T
```
```suggestion
    struct make_shared_enabler : public T
```

antonio2368 replied:
Yep, was thinking it could be related to the failures, but it actually doesn't make sense as `make_shared_enabler` isn't used anywhere directly so visibility specifier shouldn't matter here.
Yeah, let's see how many classes have `final` and are affected by this.

alexey-milovidov commented:
What about removing "shared_ptr_helper" as a whole? Looks like unreasonable complication.

rschu1ze commented:
> What about removing "shared_ptr_helper" as a whole? Looks like unreasonable complication.

That sounds reasonable to me. The only code encapsulated by `shared_ptr_helper` is a simple one-line `create()` method, and there is not really something that can wrong if `create()` was implemented directly.

In the meantime, I addressed the compilation failure in the original commit by using the "passkey" idiom in `shared_ptr_helper` instead of inheritance-from-a-helper. For that, ca. 120 classes which are currently inheriting from `shared_ptr_helper` had to be modified. In more detail, the current be`friend`-ing of the `shared_ptr_helper` in the using classes had to be replaced by a new public constructor, leading to a large commit with +800/-400 LOC. I have not pushed this change yet and the additional constructor is admittedly quite ugly as well.

So yes, removing `shared_ptr_helper` seems the best option at this point. As a newcomer, I am actually a bit afraid of annoying the developer who originally added `shared_ptr_helper` with a removal. According to Git history, the helper was added prior to 2016. So if Alexey gives his go I suppose we are good 😄  Let me try to get rid of the helper then.

Commit:
Remove shared_ptr_helper

- The logic encapsulated in "shared_ptr_helper" was a trivial, i.e. just
  a oneliner. Removed it as per discussion in the PR.

- Replaced the current machinery for shared_ptr_helper (inheritance +
  friend declaration) by a direct implementation of "create()" in the
  user classes, this time based on make_shared<>() instead of
  shared_ptr<>() to save one heap allocation.

- To disallow bypassing T::create(), i.e. direct instantiation of class
  T, the constructors of T were in almost all cases declared private or
  protected. However, unlike shared_ptr<T>(), make_shared<T>() needs
  access to a public constructor of T. I therefore applied the passkey
  idiom which adds a public forwarding constructor that can only be
  called with a 'passkey' private to T. ( https://stackoverflow.com/a/8147326).
  As a result, we get exactly the same behavior as before.

- There were a handful of classes inheriting from shared_ptr_helper
  which had a public constructor, e.g. StorageMySQL. I assume this was
  unintentional as it allows direct construction. I made these
  constructors private too.

- Also moved the shared_ptr traits to the only place where they are
  actually needed (typeid_cast.h) to reduce include bloat.

Commit:
Merge remote-tracking branch 'origin/master' into shared_ptr_helper

Commit:
Merge remote-tracking branch 'origin/master' into shared_ptr_helper

Commit:
Fix build (hopefully)

- Clang/GCC: StorageMaterializedMySQL's ctor was made private
  --> using create() now

- GCC: [[maybe_unused]] was rejected for not-very-clear reasons
  --> commented two members variables out

rschu1ze commented:
The change (dc0df67548319a) got a bit larger than expected :smiley:

Interestingly, my local build of dc0df67548319a succeeded (mac + clang) whereas the CI builds failed with a compile error in the MySQL storage code. This code is enabled/diabled by CMake option -DENABLE_MYSQL. My local CMake invocation is pretty minimal (just specifying where to find the compiler/linker + the "debug" as build type). Since ENABLE_MYSQL is also not activated "magically" within the build scripts (e.g. depending on the build type) MySQL support was disabled when compiling locally, just like expected. However, the CMake invocation in the [CI build e.g. BuilderDebDebug](https://github.com/ClickHouse/ClickHouse/runs/6126205708?check_suite_focus=true) is

`cmake --debug-trycompile --verbose=1 -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=Debug -DSANITIZE= -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_TESTS=0 -DENABLE_UTILS=0 -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_AUTOGEN_VERBOSE=ON -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DBUILD_STANDALONE_KEEPER=ON -DCMAKE_C_COMPILER=clang-13 -DCMAKE_CXX_COMPILER=clang++-13 -DENABLE_TESTS=1 ..`

which is also not enabling MySQL. Hmm, there seems to be some magic going in the CI builds which I don't understand yet. Anyways, I have pushed a compile fix.

Commit:
Fix code formatting

antonio2368 commented:
Hmm this requires now a bunch of stuff to be added to make it work (PassKey + extra constructor).
AFAIK, almost all of the objects are Storages/Database and I wouldn't say that they require such an optimisation (not on hot path) so I would honestly just keep it simple and create `std::shared_ptr` using the constructor (while still removing `shared_ptr_helper`.
@alexey-milovidov could you help me please with the decision here?

rschu1ze commented:
I agree about the negligible performance gains when object creation is not in a hot path. I admit I wasn't aware about the hot path when I started this PR. (There were some edge cases where make_shared() was [exception safe](https://stackoverflow.com/a/17355375) and shared_ptr() wasn't but these were fixed with C++17).

Yes, the passkey struct + extra constructor adds some code and is not pretty. The only reason for this machinery was to disallow direct creation of the object, i.e. to replicate the previous behavior. In more detail: all constructors are private but we still allow make_shared (which needs a public ctor) to create the object using the public passkey constructor.

I think the original author of shared_ptr_helper was afraid that someone could copy a storage object. E.g.

```
StorageS3 storageS3_1;
StorageS3 storageS3_2 = storageS3_1;
```

To me, such a (deep) copy seem unintended, maybe even dangerous, and make no sense? Perhaps someone a dev with more experience may comment. If the storage object is wrapped as shared_ptr, this situation cannot happen, at least not easily.

Anyways, I also for simplicity. To prevent copying and to simplify, how about this proposal:
1. we get rid of `shared_ptr_helper` base class +
2. instead, we add to each class a `create()`method returning a `shared_ptr` +
3. we get rid of the passkey struct and the passkey constructor and we make all normal constructors public (so make_shared in create() can call them) +
4. we inherit from boost::noncopyable to disallow copying (as now direct construction is allowed due to public constructors)

I have not tried it out yet but this should simplify the code and still be safe. Boost::noncopyable is already used a lot in ClickHouse and well known.

rschu1ze commented:
True 😉

antonio2368 commented:
So here are my thoughts.
I would 
 - make all the constructors public that are used outside of the class
 - remove all the create functions and simply call `std::make_shared` instead
 - make all those classes inherit from `boost::noncopyable` because I have a feeling that the author was aiming at that
 
 What's funny is that the classes could still call copy/move constructor (if it's not implicitly deleted) because no one forbids you to dereference `std::shared_ptr` you got.

@rschu1ze what do you think?

rschu1ze commented:
Sounds good to me.

I'll try to make some time to work on this today. In the meantime, allow me to convert this PR into a draft so that it disappears from the "ready-for-review/merge" list.

Commit:
Abolish passkey idiom and make classes noncopyable

As discussed in the PR, this commit
- makes all ctors used by outside callers public
- removes existing create() calls in favor of make_shared() on the end
  of the creator
- inherit from boost::noncopyable to prevent (potentially dangerous)
  copying of storage objects

rschu1ze commented:
That was another big commit but it Antonio's proposal worked nicely and it (imho) improves the overall state of affairs.

Commit:
Fix build

rschu1ze commented:
I joined ClickHouse Inc. as per today. This means that my mail address changed from `rs.github@mailbox.org` to `robert@clickhouse.com`. The existing commits in this PR still reference the old mail address, and since I already (perhaps too eagerly) deleted the GPG key associated with the old mail address, they all show as  "unverified" now.

As far as I know, only commits with "verified" commiters are eligible for merge. I was able to fix other PRs stuck with the same situation by squashing and force-pushing everything. In this PR, I have merged from master multiple times and I am not sure if the same trick would work.

I am therefore closing this PR and will continue this work in a fresh PR. The squash should also clean up the messy commit history :D Thanks for the review/comments so far.

alexey-milovidov commented:
> As far as I know, only commits with "verified" commiters are eligible for merge.

There is no such requirement, we can merge existing PRs with commits from the old address or even with commits from both addresses.

rschu1ze commented:
> There is no such requirement, we can merge existing PRs with commits from the old address or even with commits from both addresses.

Alright, I'll next time. The continuation of this PR is here: #36844.
